File: Question/17_Letter_Combinations_of_a_Phone_Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if(digits.empty()) return {};
        vector<string> letters = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string> ans{""};
        for(int i = 0; i < digits.size(); ++i) {
            string letter = letters[digits[i] - '0'];
            vector<string> t;
            for(auto& ele : letter) {
                for(auto& k : ans)
                    t.push_back(k + ele);
            }
            ans = t;
        }
        return ans;
    }
};

//dfs
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if(digits.empty()) return {};
        vector<string> letters = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string> ans;
        dfs(digits, letters, ans, "", 0);
        return ans;
    }
    void dfs(string& digits, vector<string>& letters, vector<string>& ans, string cur, int level) {
        if(level == digits.size()) {
            ans.push_back(cur);
            return;
        }
        string str = letters[digits[level] - '0'];
        for(int i = 0; i < str.size(); ++i) {
            dfs(digits, letters, ans, cur + str[i], level + 1);
        }
    }
};

File: Question_Java/17_Letter_Combinations_of_a_Phone_Number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits.isEmpty())
            return new ArrayList<String>();
        String[] dict = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
        char[] chs = digits.toCharArray();
        List<String> ans = new ArrayList<>();
        dfs(chs, dict, 0, new StringBuilder(), ans);
        return ans;
    }

    void dfs(char[] chs, String[] dict, int level, StringBuilder cur, List<String> ans) {
        if (level == chs.length) {
            ans.add(cur.toString());
            return;
        }

        int num = chs[level] - '0';
        for (int i = 0; i < dict[num].length(); ++i) {
            cur.append(dict[num].charAt(i));
            dfs(chs, dict, level + 1, cur, ans);
            cur.deleteCharAt(cur.length() - 1);
        }
    }
}